pkg/actors/internal/client.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+	"sync"
+
+	v1pb "github.com/dapr/dapr/pkg/proto/placement/v1"
+
+	"google.golang.org/grpc"
+)
+
+// placementClient implements the best practices when handling grpc streams
+// using exclusion locks to prevent concurrent Recv and Send from channels.
+// properly handle stream closing by draining the renamining events until receives an error.
+// and broadcasts connection results based on new connections and disconnects.
+type placementClient struct {
+	// getGrpcOpts are the options that should be used to connect to the placement service
+	getGrpcOpts func() ([]grpc.DialOption, error)
+
+	// recvLock prevents concurrent access to the Recv() method.
+	recvLock *sync.Mutex
+	// sendLock prevents CloseSend and Send being used concurrently.
+	sendLock *sync.Mutex
+
+	// clientStream is the client side stream.
+	clientStream v1pb.Placement_ReportDaprStatusClient
+	// clientConn is the gRPC client connection.
+	clientConn *grpc.ClientConn
+	// streamConnAlive is the status of stream connection alive.
+	streamConnAlive bool
+	// streamConnectedCond is the condition variable for goroutines waiting for or announcing
+	// that the stream between runtime and placement is connected.
+	streamConnectedCond *sync.Cond
+
+	// ctx is the stream context
+	ctx context.Context
+	// cancel is the cancel func for context
+	cancel context.CancelFunc
+}
+
+// connectToServer initializes a new connection to the target server and if it succeeds replace the current
+// stream with the connected stream.
+func (c *placementClient) connectToServer(serverAddr string) error {
+	opts, err := c.getGrpcOpts()
+	if err != nil {
+		return err
+	}
+
+	conn, err := grpc.Dial(serverAddr, opts...)
+	if err != nil {
+		if conn != nil {
+			conn.Close()
+		}
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Background should be the way to go for the context of this PR, because it has been used since the first file version, so I’d rather prefer keep using background even if there’s alternatives then I can check it later on, how that’s sound?

Reply:
OK